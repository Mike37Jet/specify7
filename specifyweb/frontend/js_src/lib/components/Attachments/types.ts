import { RA } from '../../utils/types';
import { SpecifyResource } from '../DataModel/legacyTypes';
import { Attachment } from '../DataModel/types';

export type UploadAttachmentSpec = {
  readonly token: string;
  readonly attachmentlocation: string;
};

export type AttachmentStatus =
  | 'incorrectFormatter'
  | 'noFile'
  | 'uploaded'
  | 'deleted'
  | { readonly type: 'skipped' | 'cancelled'; readonly reason: string };

export type PartialUploadableFileSpec = Partial<UploadableFileSpec> &
  Pick<UploadableFileSpec, 'file'>;

type UploadableFileSpec = {
  readonly file: UnBoundFile;
  readonly matchedId: RA<number>;
  readonly status: AttachmentStatus;
  readonly disambiguated: number;
  readonly attachmentId: number;
  readonly uploadTokenSpec: UploadAttachmentSpec;
};

export type CanUpload = Pick<
  PartialUploadableFileSpec,
  'disambiguated' | 'status' | 'uploadTokenSpec'
> &
  Omit<
    UploadableFileSpec,
    'attachmentId' | 'disambiguated' | 'status' | 'uploadTokenSpec' | 'file'
  > & { readonly file: Required<FileWithExtras> };

export type CanDelete = Omit<
  PartialUploadableFileSpec,
  'matchedId' | 'attachmentId'
> &
  Pick<UploadableFileSpec, 'matchedId' | 'attachmentId'>;

export type UploadInternalWorkable<ACTION extends 'uploading' | 'deleting'> =
  ACTION extends 'uploading'
    ? {
        readonly canUpload: true;
        readonly attachmentFromPreviousTry?: SpecifyResource<Attachment>;
      } & CanUpload
    : {
        readonly canDelete: true;
      } & CanDelete;

// Weakens types if we can't perform upload or delete. Strictens them if we can.
export type TestInternalUploadSpec<ACTION extends 'uploading' | 'deleting'> =
  ACTION extends 'uploading'
    ?
        | UploadInternalWorkable<'uploading'>
        | ({ readonly canUpload: false } & PartialUploadableFileSpec)
    :
        | UploadInternalWorkable<'deleting'>
        | ({ readonly canDelete: false } & PartialUploadableFileSpec);

// Weak type. Used because after uploading/deleting, return spec becomes fuzzy in types.
export type PostWorkUploadSpec<ACTION extends 'uploading' | 'deleting'> =
  PartialUploadableFileSpec &
    (ACTION extends 'uploading'
      ? {
          readonly canUpload: boolean;
        }
      : { readonly canDelete: boolean });

type FileWithExtras = File & {
  parsedName?: string;
};

// Forcing keys to be primitive because objects would be
// ignored during syncing to backend.
export type BoundFile = Pick<
  {
    [P in keyof FileWithExtras as FileWithExtras[P] extends object
      ? never
      : P]: FileWithExtras[P];
  },
  'size' | 'name' | 'parsedName' | 'type'
>;

export type UnBoundFile = FileWithExtras | BoundFile;
