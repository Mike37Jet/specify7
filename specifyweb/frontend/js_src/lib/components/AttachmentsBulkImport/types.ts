import type React from 'react';

import type { RA } from '../../utils/types';
import type { PartialAttachmentUploadSpec } from './Import';
import type { staticAttachmentImportPaths } from './importPaths';
import type { keyLocalizationMapAttachment } from './utils';

export type UploadAttachmentSpec = {
  readonly token: string;
  readonly attachmentLocation: string;
};
type Matched = { readonly type: 'matched'; readonly id: number };

export type AttachmentStatus =
  | Matched
  | {
      readonly type: 'cancelled' | 'skipped';
      readonly reason: keyof typeof keyLocalizationMapAttachment;
    }
  | { readonly type: 'success'; readonly successType: 'deleted' | 'uploaded' };

export type PartialUploadableFileSpec = Partial<UploadableFileSpec> &
  Pick<UploadableFileSpec, 'file'>;

type UploadableFileSpec = {
  readonly file: UnBoundFile;
  readonly matchedId: RA<number>;
  readonly status: AttachmentStatus;
  readonly disambiguated: number;
  readonly attachmentId: number;
  readonly uploadTokenSpec: UploadAttachmentSpec;
};

type FileWithExtras = File & {
  // eslint-disable-next-line functional/prefer-readonly-type
  parsedName?: string;
};

/*
 * Forcing keys to be primitive because objects would be
 * ignored during syncing to backend.
 */
export type BoundFile = Pick<
  FileWithExtras,
  'name' | 'parsedName' | 'size' | 'type'
>;

export type UnBoundFile = BoundFile | FileWithExtras;

export type AttachmentWorkProgress = {
  readonly total: number;
  readonly uploaded: number;
  readonly type: 'interrupted' | 'safe' | 'stopped' | 'stopping';
  readonly retryingIn: number;
};

export type AttachmentWorkRef = {
  // eslint-disable-next-line functional/prefer-readonly-type
  mappedFiles: RA<PartialUploadableFileSpec>;
  // eslint-disable-next-line functional/prefer-readonly-type
  uploadPromise: Promise<number | undefined>;
  readonly retrySpec: Record<number, number>;
};

export type AttachmentWorkStateProps = {
  readonly onCompletedWork: (
    uploadables: RA<PartialUploadableFileSpec> | undefined
  ) => void;
} & {
  readonly workProgress: AttachmentWorkProgress;
  readonly workRef: React.MutableRefObject<AttachmentWorkRef>;
  readonly onStop: () => void;
  readonly triggerNow: () => void;
};

type SavedDataSetFields = {
  readonly id: number;
  readonly timestampcreated: string;
  readonly timestampmodified?: string;
};

export type AttachmentDataSetResource = {
  readonly name: string;
  readonly rows: RA<PartialUploadableFileSpec>;
  readonly uploaderstatus:
    | 'deleting'
    | 'deletingInterrupted'
    | 'main'
    | 'renaming'
    | 'uploading'
    | 'uploadInterrupted'
    | 'validating';
  readonly uploadplan: PartialAttachmentUploadSpec;
};

export type SavedAttachmentDataSetResource = AttachmentDataSetResource &
  SavedDataSetFields;

export type FetchedDataSet =
  | SavedAttachmentDataSetResource &
      (
        | { readonly uploaderstatus: 'main' }
        | ({
            readonly uploaderstatus: 'deleting' | 'uploading';
            readonly rows: RA<PartialUploadableFileSpec>;
          } & {
            readonly uploadplan: {
              readonly staticPathKey: keyof typeof staticAttachmentImportPaths;
            };
          })
      );
