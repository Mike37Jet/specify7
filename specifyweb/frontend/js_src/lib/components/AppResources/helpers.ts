import type { AppResourcesConformation } from './Aside';
import type { AppResources, AppResourcesTree } from './hooks';
import type { SpAppResource, SpViewSetObj } from '../DataModel/types';
import { f } from '../../utils/functools';
import type { SpecifyModel } from '../DataModel/specifyModel';
import type { RA } from '../../utils/types';
import { SerializedResource } from '../DataModel/helperTypes';

export const getAppResource = (
  resources: AppResources,
  model: SpecifyModel<SpAppResource | SpViewSetObj>,
  resourceId: number | undefined
): SerializedResource<SpAppResource | SpViewSetObj> | false | undefined =>
  typeof resourceId === 'number'
    ? (
        (model.name === 'SpAppResource'
          ? resources.appResources
          : resources.viewSets) as RA<SerializedResource<SpAppResource>>
      ).find(({ id }) => id === resourceId) ?? false
    : undefined;

export type AppResourceMode = 'appResources' | 'viewSets';

export const getAppResourceMode = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): AppResourceMode =>
  resource._tableName === 'SpAppResource' ? 'appResources' : 'viewSets';

export const getAppResourceCount = (
  resourcesTree: AppResourcesTree[number]
): number =>
  resourcesTree.appResources.length +
  resourcesTree.viewSets.length +
  f.sum(resourcesTree.subCategories.map(getAppResourceCount));

export const buildAppResourceConformation = (
  resourcesTree: AppResourcesTree
): RA<AppResourcesConformation> =>
  resourcesTree
    .filter(
      ({ subCategories, appResources, viewSets }) =>
        subCategories.length > 0 ||
        appResources.length > 0 ||
        viewSets.length > 0
    )
    .map(({ key, subCategories }) => ({
      key,
      children: buildAppResourceConformation(subCategories),
    }));
