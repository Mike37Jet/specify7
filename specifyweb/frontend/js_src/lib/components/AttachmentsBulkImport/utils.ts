import { attachmentsText } from '../../localization/attachments';
import { commonText } from '../../localization/common';
import { wbText } from '../../localization/workbench';
import { ajax } from '../../utils/ajax';
import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { filterArray } from '../../utils/types';
import {
  insertItem,
  keysToLowerCase,
  mappedFind,
  replaceItem,
} from '../../utils/utils';
import { addMissingFields } from '../DataModel/addMissingFields';
import { deserializeResource, serializeResource } from '../DataModel/helpers';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { schema } from '../DataModel/schema';
import type { SpQuery, SpQueryField, Tables } from '../DataModel/types';
import type { UiFormatter } from '../Forms/uiFormatters';
import { queryFieldFilters } from '../QueryBuilder/FieldFilter';
import { makeQueryField } from '../QueryBuilder/fromTree';
import type { AttachmentUploadSpec } from './Import';
import { AttachmentMapping, staticAttachmentImportPaths } from './importPaths';
import type {
  AttachmentStatus,
  CanDelete,
  CanUpload,
  PartialUploadableFileSpec,
  UnBoundFile,
} from './types';
import { formatterTypeMapper } from '../Forms/uiFormatters';
import { State } from 'typesafe-reducer';
import { formsText } from '../../localization/forms';

const isAttachmentMatchValid = (uploadSpec: PartialUploadableFileSpec) =>
  uploadSpec.matchedId !== undefined &&
  (uploadSpec.matchedId.length === 1 || uploadSpec.disambiguated !== undefined);

export type ResolvedAttachmentRecord =
  | State<
      'invalid' | 'valid',
      { readonly reason: keyof typeof keyLocalizationMapAttachment }
    >
  | State<'matched', { readonly id: number }>;

const resolveAttachmentMatch = (
  matchedId: RA<number>,
  disambiguated: number | undefined
): ResolvedAttachmentRecord =>
  matchedId.length === 0
    ? { type: 'invalid', reason: 'noMatch' }
    : matchedId.length > 1 && disambiguated === undefined
    ? {
        type: 'invalid',
        reason: 'multipleMatches',
      }
    : {
        type: 'matched',
        id: disambiguated ?? matchedId[0],
      };
export function resolveAttachmentRecord(
  matchedId: RA<number> | undefined,
  disambiguated: number | undefined,
  parsedName: string | undefined
): ResolvedAttachmentRecord {
  if (parsedName === undefined)
    return { type: 'invalid', reason: 'incorrectFormatter' };
  if (matchedId === undefined)
    return { type: 'valid', reason: 'correctlyFormatted' };
  return resolveAttachmentMatch(matchedId, disambiguated);
}

const findFirstReason =
  (
    conditions: (uploadSpec: PartialUploadableFileSpec) => RA<boolean>,
    humanReasons: RA<keyof typeof keyLocalizationMapAttachment>
  ) =>
  (uploadSpec: PartialUploadableFileSpec) => {
    const resolvedConditions = conditions(uploadSpec);
    return mappedFind(resolvedConditions, (condition, index) =>
      condition ? undefined : humanReasons[index]
    );
  };

const uploadHumanReasons: RA<keyof typeof keyLocalizationMapAttachment> = [
  'alreadyUploaded',
  'noFile',
  'incorrectFormatter',
  'matchError',
];

const deleteHumanReasons: RA<keyof typeof keyLocalizationMapAttachment> = [
  'noAttachments',
  'matchError',
];
export const reasonToSkipUpload = findFirstReason(
  (uploadSpec) => [
    uploadSpec.attachmentId === undefined,
    uploadSpec.file instanceof File,
    uploadSpec.file.parsedName !== undefined,
    isAttachmentMatchValid(uploadSpec),
  ],
  uploadHumanReasons
);

export const reasonToSkipDelete = findFirstReason(
  (uploadSpec) => [
    uploadSpec.attachmentId !== undefined,
    isAttachmentMatchValid(uploadSpec),
  ],
  deleteHumanReasons
);

export const canUploadAttachment = (
  uploadSpec: PartialUploadableFileSpec
): uploadSpec is CanUpload => reasonToSkipUpload(uploadSpec) === undefined;

export const canDeleteAttachment = (
  uploadSpec: PartialUploadableFileSpec
): uploadSpec is CanDelete =>
  uploadSpec.attachmentId !== undefined && isAttachmentMatchValid(uploadSpec);

function generateInQueryResource(
  baseTable: keyof Tables,
  path: string,
  searchableList: RA<number | string | undefined>,
  queryName: string,
  additionalPaths: RA<
    Partial<SerializedResource<SpQueryField>> & { readonly path: string }
  > = []
): SpecifyResource<SpQuery> {
  const rawFields = [
    ...additionalPaths,
    {
      path,
      isDisplay: true,
      operStart: queryFieldFilters.in.id,
      startValue: filterArray(searchableList).join(','),
    },
  ];
  const queryField = rawFields.map(({ path, ...field }, index) =>
    serializeResource(
      makeQueryField(baseTable, path, { ...field, position: index })
    )
  );
  return deserializeResource(
    addMissingFields('SpQuery', {
      name: queryName,
      contextName: baseTable,
      contextTableId: schema.models[baseTable].tableId,
      countOnly: false,
      fields: queryField,
    })
  );
}

export async function validateAttachmentFiles(
  uploadableFiles: RA<PartialUploadableFileSpec>,
  uploadSpec: AttachmentUploadSpec,
  keepDisambiguation: boolean = false
): Promise<RA<PartialUploadableFileSpec>> {
  const { baseTable, path } =
    staticAttachmentImportPaths[uploadSpec.staticPathKey];

  const validationQueryResource = generateInQueryResource(
    baseTable,
    path,
    uploadableFiles.map(({ file }) => file?.parsedName),
    'Batch Attachment Upload'
  );
  const validationResponse = (
    await validationPromiseGenerator(validationQueryResource)
  )
    .map(({ targetId, restResult }) => ({
      targetId,
      rawResult: uploadSpec.formatQueryResults(restResult[0]),
    }))
    .filter((result) => result.rawResult !== undefined);

  return matchFileSpec(
    uploadableFiles,
    validationResponse as RA<{
      readonly targetId: number;
      readonly rawResult: string;
    }>,
    keepDisambiguation
  );
}

export function matchSelectedFiles(
  previousUploadables: RA<PartialUploadableFileSpec>,
  filesToResolve: RA<PartialUploadableFileSpec>
): RA<PartialUploadableFileSpec> {
  let resolvedFiles = previousUploadables;
  filesToResolve.forEach((uploadable) => {
    const matchedIndex = resolvedFiles.findIndex((previousUploadable) => {
      if (
        previousUploadable.attachmentId !== undefined ||
        previousUploadable.file instanceof File
      )
        return false;

      return (
        previousUploadable.file !== undefined &&
        previousUploadable.file.name === uploadable.file.name &&
        previousUploadable.file.size === uploadable.file.size &&
        previousUploadable.file.type === uploadable.file.type
      );
    });

    resolvedFiles =
      matchedIndex === -1
        ? insertItem(resolvedFiles, resolvedFiles.length, uploadable)
        : replaceItem(resolvedFiles, matchedIndex, {
            ...resolvedFiles[matchedIndex],
            file: uploadable.file,
            /*
             * Generating tokens again because the file could have been
             * uploaded to the asset server but not yet recorded in Specify DB.
             */
            uploadTokenSpec: undefined,
            // Take the new status in case of parse failure was reported.
            status: uploadable.status,
          });
  });
  return resolvedFiles;
}

export function resolveFileNames(
  previousFile: UnBoundFile,
  getFormatted: (
    rawName: number | string | null | undefined
  ) => string | undefined,
  formatter?: UiFormatter
): PartialUploadableFileSpec {
  let nameToParse: string | undefined;
  if (
    formatter === undefined ||
    formatter.fields.some((field) => field instanceof formatterTypeMapper.regex)
  )
    nameToParse = stripLastOccurrence(previousFile.name, '.');
  else {
    const formattedLength = formatter.fields.reduce(
      (length, field) => length + field.size,
      0
    );
    nameToParse = previousFile.name.slice(0, formattedLength);
  }
  previousFile.parsedName = f.maybe(nameToParse, getFormatted);

  return {
    file: previousFile,
  };
}

export function stripLastOccurrence(target: string, delimiter: string) {
  const splittedString = target.split(delimiter);
  return splittedString
    .slice(0, splittedString.length === 1 ? 1 : -1)
    .join(delimiter);
}

const validationPromiseGenerator = async (
  queryResource: SpecifyResource<SpQuery>
): Promise<
  RA<{
    readonly targetId: number;
    readonly restResult: RA<number | string>;
  }>
> =>
  ajax<{
    // First value is the primary key
    readonly results: RA<readonly [number, ...RA<number | string>]>;
  }>('/stored_query/ephemeral/', {
    method: 'POST',
    headers: {
      // eslint-disable-next-line @typescript-eslint/naming-convention
      Accept: 'application/json',
    },
    body: keysToLowerCase({
      ...serializeResource(queryResource),
      countOnly: false,
      limit: 0,
    }),
  }).then(({ data }) =>
    data.results.map(([target, ...restResult]) => ({
      targetId: target,
      restResult,
    }))
  );

const matchFileSpec = (
  uploadFileSpec: RA<PartialUploadableFileSpec>,
  queryResults: RA<{
    readonly targetId: number;
    readonly rawResult: string;
  }>,
  keepDisambiguation: boolean = false
): RA<PartialUploadableFileSpec> =>
  uploadFileSpec.map((spec) => {
    const specParsedName = spec.file?.parsedName;
    // Don't match files already uploaded.
    if (specParsedName === undefined || typeof spec.attachmentId === 'number')
      return spec;
    const matchingResults = queryResults.filter(
      (result) => result.rawResult === specParsedName
    );
    const newSpec: PartialUploadableFileSpec = {
      ...spec,
      matchedId: matchingResults.map((result) => result.targetId),
    };

    if (
      keepDisambiguation &&
      spec.disambiguated !== undefined &&
      // If disambiguation was chosen, but it became invalid, reset disambiguation
      newSpec.matchedId?.includes(spec.disambiguated)
    ) {
      return newSpec;
    }
    return { ...newSpec, disambiguated: undefined };
  });

export async function reconstructDeletingAttachment(
  staticKey: keyof typeof staticAttachmentImportPaths,
  deletableFiles: RA<PartialUploadableFileSpec>
): Promise<RA<PartialUploadableFileSpec>> {
  const baseTable = staticAttachmentImportPaths[staticKey].baseTable;
  const path = `${AttachmentMapping[baseTable].relationship}.${staticAttachmentImportPaths[staticKey].restPath}`;
  const relatedAttachments = filterArray(
    deletableFiles.map((deletable) =>
      deletable.status?.type === 'matched' ? deletable.attachmentId : undefined
    )
  );
  const reconstructingQueryResource = generateInQueryResource(
    baseTable,
    path,
    relatedAttachments,
    'Batch Attachment Upload'
  );
  const queryResults = await validationPromiseGenerator(
    reconstructingQueryResource
  );
  return deletableFiles.map((deletable) => {
    if (deletable.status?.type !== 'matched') return deletable;
    const matchedId = deletable.status.id;
    const foundInQueryResult = queryResults.find(
      ({ targetId, restResult: [attachmentId] }) =>
        targetId === matchedId && attachmentId === deletable.attachmentId
    );
    return {
      ...deletable,
      attachmentId: foundInQueryResult?.restResult[0] as number,
      status:
        foundInQueryResult === undefined
          ? ({ type: 'success', successType: 'uploaded' } as const)
          : deletable.status.type === 'matched'
          ? ({
              type: 'cancelled',
              reason: 'rollbackInterruption',
            } as const)
          : deletable.status,
    };
  });
}
export async function reconstructUploadingAttachmentSpec(
  staticKey: keyof typeof staticAttachmentImportPaths,
  uploadableFiles: RA<PartialUploadableFileSpec>
): Promise<RA<PartialUploadableFileSpec>> {
  const baseTable = staticAttachmentImportPaths[staticKey].baseTable;
  const pathToAttachmentLocation = `${AttachmentMapping[baseTable].relationship}.attachment.attachmentLocation`;
  const filteredAttachmentLocations = filterArray(
    uploadableFiles.map((uploadable) =>
      uploadable.status?.type === 'matched'
        ? uploadable.uploadTokenSpec?.attachmentlocation
        : undefined
    )
  );
  const reconstructingQueryResource = generateInQueryResource(
    baseTable,
    pathToAttachmentLocation,
    filteredAttachmentLocations,
    'Batch Attachment Upload',
    [
      {
        path: `${AttachmentMapping[baseTable].relationship}.${staticAttachmentImportPaths[staticKey].restPath}`,
        isDisplay: true,
        id: queryFieldFilters.any.id,
      },
    ]
  );
  const queryResults = await validationPromiseGenerator(
    reconstructingQueryResource
  );
  return uploadableFiles.map((uploadable) => {
    if (uploadable.status?.type !== 'matched') return uploadable;
    const matchedId = uploadable.status.id;
    const foundInQueryResult = queryResults.find(
      ({ targetId, restResult: [_, attachmentLocation] }) =>
        typeof attachmentLocation === 'string' &&
        targetId === matchedId &&
        attachmentLocation.toString() ===
          uploadable.uploadTokenSpec!.attachmentlocation
    );

    return {
      ...uploadable,
      attachmentId: foundInQueryResult?.restResult[0] as number,
      status:
        typeof foundInQueryResult === 'object'
          ? ({ type: 'success', successType: 'uploaded' } as const)
          : uploadable.status.type === 'matched'
          ? ({
              type: 'cancelled',
              reason: 'uploadInterruption',
            } as const)
          : uploadable.status,
    };
  });
}

export const keyLocalizationMapAttachment = {
  incorrectFormatter: attachmentsText.incorrectFormatter(),
  noFile: attachmentsText.noFile(),
  uploaded: commonText.uploaded(),
  deleted: attachmentsText.deleted(),
  alreadyUploaded: attachmentsText.alreadyUploaded(),
  alreadyDeleted: attachmentsText.alreadyDeleted(),
  skipped: attachmentsText.skipped(),
  cancelled: attachmentsText.cancelled(),
  matchError: attachmentsText.matchError(),
  noAttachments: attachmentsText.noAttachments(),
  uploadInterruption: attachmentsText.frontEndInterruption({
    action: wbText.upload(),
  }),
  rollbackInterruption: attachmentsText.frontEndInterruption({
    action: wbText.rollback(),
  }),
  errorReadingFile: attachmentsText.errorReadingFile(),
  attachmentServerUnavailable: attachmentsText.attachmentServerUnavailable(),
  saveConflict: formsText.saveConflict(),
  unhandledFatalResourceError: attachmentsText.unhandledFatalResourceError(),
  nothingFound: formsText.nothingFound(),
  noMatch: attachmentsText.noMatch(),
  multipleMatches: attachmentsText.multipleMatches(),
  correctlyFormatted: attachmentsText.correctlyFormatted(),
  userStopped: attachmentsText.stoppedByUser(),
} as const;

export function resolveAttachmentStatus(
  attachmentStatus: AttachmentStatus
): string {
  if ('reason' in attachmentStatus) {
    const reason = keyLocalizationMapAttachment[attachmentStatus.reason];
    return commonText.colonLine({
      label: keyLocalizationMapAttachment[attachmentStatus.type],
      value: reason,
    });
  }
  return attachmentStatus.type === 'success'
    ? keyLocalizationMapAttachment[attachmentStatus.successType]
    : '';
}
