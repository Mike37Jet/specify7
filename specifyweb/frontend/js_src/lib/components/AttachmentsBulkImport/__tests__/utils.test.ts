import { requireContext } from '../../../tests/helpers';
import {
  CatalogNumberNumeric,
  formatterTypeMapper,
  UiFormatter,
} from '../../Forms/uiFormatters';
import { IR, RA } from '../../../utils/types';
import { schema } from '../../DataModel/schema';
import { formatterToParser } from '../../../utils/parser/definitions';
import { syncFieldFormat } from '../../../utils/fieldFormat';
import { matchFileSpec, resolveFileNames } from '../utils';
import { PartialUploadableFileSpec } from '../types';

requireContext();

type TestDefinition = IR<{
  readonly uiFormatter: UiFormatter;
  readonly testCases: RA<[string, undefined | string]>;
}>;

const fileNameTestSpec: TestDefinition = {
  catalogNumber: {
    uiFormatter: new CatalogNumberNumeric(),
    testCases: [
      ['000000001.jpg', '000000001'],
      ['00001.jpg', '000000001'],
      ['00001(1).jpg', undefined],
      ['000000001(2).jpg', '000000001'],
      // BUG: This could lead to unexpected matches
      ['0000000000002.jpg', '000000000'],
      ['someRandomValue.jpg', undefined],
    ],
  },
  numeric: {
    uiFormatter: new UiFormatter(false, [
      new formatterTypeMapper.numeric({
        size: 3,
        autoIncrement: true,
        byYear: false,
      }),
    ]),
    testCases: [
      ['000.jpg', '000'],
      ['001.jpg', '001'],
      ['23.jpg', '023'],
      ['DEF001.jpg', undefined],
      ['000(1).jpg', '000'],
      ['23(1).jpg', undefined],
    ],
  },
  regex: {
    uiFormatter: new UiFormatter(false, [
      new formatterTypeMapper.regex({
        size: 3,
        autoIncrement: true,
        value: '^\\d{1,6}(?:[a-zA-Z]{1,2})?$',
        byYear: false,
      }),
    ]),
    testCases: [
      ['45265.jpg', '45265'],
      ['45622AB.jpg', '45622AB'],
      ['45622AB', '45622AB'],
      ['45622787ABCDEF.jpg', undefined],
    ],
  },
};

describe('file names resolution test', () => {
  // Using text1 to not confuse with potential catalog number autonumbering
  Object.entries(fileNameTestSpec).forEach(
    ([testName, { uiFormatter, testCases }]) => {
      test(testName, () => {
        jest.spyOn(console, 'error').mockImplementation();
        const field = schema.models.CollectionObject.getLiteralField('text1')!;
        const getResultFormatter = (formatter: UiFormatter) => {
          return (value: string | undefined | number) =>
            value === undefined || value === null
              ? undefined
              : syncFieldFormat(
                  field,
                  formatterToParser(field, formatter),
                  value.toString(),
                  true
                );
        };
        const resultFormatter = getResultFormatter(uiFormatter);
        testCases.forEach(([input, output]) =>
          expect(resolveFileNames(input, resultFormatter, uiFormatter)).toEqual(
            output
          )
        );
      });
    }
  );
});

describe('resolve file names', () => {
  const queryResults = [
    [0, '1'] as const,
    [1, '2'] as const,
    [2, '3'] as const,
    [3, '4'] as const,
    [10, '5'] as const,
    [11, '5'] as const,
  ];
  const uploadSpec: RA<PartialUploadableFileSpec> = [
    {
      uploadFile: {
        file: { name: 'test', size: 0, type: '0' },
        parsedName: '1',
      },
    },
    {
      uploadFile: {
        file: { name: 'test2', size: 0, type: '0' },
        parsedName: '2',
      },
    },
    {
      uploadFile: {
        file: { name: 'test3', size: 0, type: '0' },
        parsedName: '3',
      },
    },
    {
      uploadFile: {
        file: { name: 'test4', size: 0, type: '0' },
        parsedName: '4',
      },
      attachmentId: 10,
      matchedId: [2],
      disambiguated: 10,
    },
    {
      // Shouldn't assume that backend will give back all the results back
      // just more than once, since MySQL will remove duplicates when having the in (..)
      // clause
      uploadFile: {
        file: { name: 'test5', size: 0, type: '0' },
        parsedName: '4',
      },
    },
    {
      uploadFile: {
        file: { name: 'test6', size: 0, type: '0' },
        parsedName: '5',
      },
      // This will be reset
      matchedId: [12, 13, 10, 11],
      disambiguated: 10,
    },
    {
      uploadFile: {
        file: { name: 'test6', size: 0, type: '0' },
        parsedName: '5',
      },
      matchedId: [10, 11],
      // This should be reset, even if keeping disambiguation since it is stale
      disambiguated: 18,
    },
    {
      uploadFile: {
        file: { name: 'test7', size: 0, type: '0' },
        parsedName: '5',
      },
      matchedId: [10, 11],
      // This should be preserved
      disambiguated: 10,
    },
  ];
  test('resolve file names from validation', () => {
    expect(matchFileSpec(uploadSpec, queryResults)).toMatchSnapshot();
  });
  test('resolve file names from validation (keeping disambiguation)', () => {
    expect(matchFileSpec(uploadSpec, queryResults, true)).toMatchSnapshot();
  });
});
