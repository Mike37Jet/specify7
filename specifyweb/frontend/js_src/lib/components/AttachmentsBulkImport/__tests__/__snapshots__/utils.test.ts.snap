// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve file names resolve file names from validation (keeping disambiguation) 1`] = `
[
  {
    "disambiguated": undefined,
    "matchedId": [
      0,
    ],
    "uploadFile": {
      "file": {
        "name": "test",
        "size": 0,
        "type": "0",
      },
      "parsedName": "1",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      1,
    ],
    "uploadFile": {
      "file": {
        "name": "test2",
        "size": 0,
        "type": "0",
      },
      "parsedName": "2",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      2,
    ],
    "uploadFile": {
      "file": {
        "name": "test3",
        "size": 0,
        "type": "0",
      },
      "parsedName": "3",
    },
  },
  {
    "attachmentId": 10,
    "disambiguated": 10,
    "matchedId": [
      2,
    ],
    "uploadFile": {
      "file": {
        "name": "test4",
        "size": 0,
        "type": "0",
      },
      "parsedName": "4",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      3,
    ],
    "uploadFile": {
      "file": {
        "name": "test5",
        "size": 0,
        "type": "0",
      },
      "parsedName": "4",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test6",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test6",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
  {
    "disambiguated": 10,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test7",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
]
`;

exports[`resolve file names resolve file names from validation 1`] = `
[
  {
    "disambiguated": undefined,
    "matchedId": [
      0,
    ],
    "uploadFile": {
      "file": {
        "name": "test",
        "size": 0,
        "type": "0",
      },
      "parsedName": "1",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      1,
    ],
    "uploadFile": {
      "file": {
        "name": "test2",
        "size": 0,
        "type": "0",
      },
      "parsedName": "2",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      2,
    ],
    "uploadFile": {
      "file": {
        "name": "test3",
        "size": 0,
        "type": "0",
      },
      "parsedName": "3",
    },
  },
  {
    "attachmentId": 10,
    "disambiguated": 10,
    "matchedId": [
      2,
    ],
    "uploadFile": {
      "file": {
        "name": "test4",
        "size": 0,
        "type": "0",
      },
      "parsedName": "4",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      3,
    ],
    "uploadFile": {
      "file": {
        "name": "test5",
        "size": 0,
        "type": "0",
      },
      "parsedName": "4",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test6",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test6",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test7",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
]
`;
